test/Facilitator/confirmRedeemIntent.js

```diff
@@ -0,0 +1,541 @@
+const SpyAssert = require('../../test_utils/SpyAssert');
+const Message = require('../../src/utils/Message');
+
+const MessageStatus = Message.messageStatus();
+const assert = chai.assert;
+
+describe('Facilitator.confirmRedeemIntent()', () => {
+  let mosaic;
+  let redeemParams;
+  let txOptions;
+  let facilitator;
+
+  let getRedeemMessageHashResult;
+  let getOutboxMessageStatusResult;
+  let getInboxMessageStatusResult;
+  let getCoGatewayProofResult;
+  let proveGatewayResult;
+  let confirmRedeemIntentResult;
+
+  let spyGetRedeemMessageHash;
+  let spyGetOutboxMessageStatus;
+  let spyGetInboxMessageStatus;
+  let spyGetCoGatewayProof;
+  let spyProveGateway;
+  let spyConfirmRedeemIntent;
+  let spyCall;
+
+  const setup = () => {
+    spyGetRedeemMessageHash = sinon.replace(
+      Message,
+      'getRedeemMessageHash',
+      sinon.fake.returns(getRedeemMessageHashResult),
+    );
+    spyGetOutboxMessageStatus = sinon.replace(
+      facilitator.coGateway,
+      'getOutboxMessageStatus',
+      sinon.fake.resolves(getOutboxMessageStatusResult),
+    );
+    spyGetInboxMessageStatus = sinon.replace(
+      facilitator.gateway,
+      'getInboxMessageStatus',
+      sinon.fake.resolves(getInboxMessageStatusResult),
+    );
+    spyGetCoGatewayProof = sinon.replace(
+      facilitator,
+      'getCoGatewayProof',
+      sinon.fake.resolves(getCoGatewayProofResult),
+    );
+    spyProveGateway = sinon.replace(
+      facilitator.gateway,
+      'proveGateway',
+      sinon.fake.resolves(proveGatewayResult),
+    );
+    spyConfirmRedeemIntent = sinon.replace(
+      facilitator.gateway,
+      'confirmRedeemIntent',
+      sinon.fake.resolves(confirmRedeemIntentResult),
+    );
+    spyCall = sinon.spy(facilitator, 'confirmRedeemIntent');
+  };
+
+  const teardown = () => {
+    spyCall.restore();
+    sinon.restore();
+  };
+
+  beforeEach(() => {
+    mosaic = TestMosaic.mosaic();
+
+    redeemParams = {
+      redeemer: '0x0000000000000000000000000000000000000001',
+      amount: '1000000',
+      beneficiary: '0x0000000000000000000000000000000000000002',
+      gasPrice: '1',
+      gasLimit: '100000000',
+      nonce: '1',
+      hashLock:
+        '0x0000000000000000000000000000000000000000000000000000000000000001',
+    };
+    txOptions = {
+      from: redeemParams.redeemer,
+      gas: '7500000',
+    };
+
+    facilitator = new Facilitator(mosaic);
+
+    getRedeemMessageHashResult =
+      '0x0000000000000000000000000000000000000000000000000000000000000002';
+    getOutboxMessageStatusResult = MessageStatus.DECLARED;
+    getInboxMessageStatusResult = MessageStatus.UNDECLARED;
+    getCoGatewayProofResult = {
+      blockNumber: '10000',
+      accountData: '0x23232323232',
+      accountProof: '0x2323232323323223',
+    };
+    proveGatewayResult = true;
+    confirmRedeemIntentResult = true;
+  });
+
+  it('should throw an error when redeemer address is undefined', async () => {
```

Comment:
These validations are already done in contract interacts. This looks like a duplicate logic in the facilitator.

Reply:
I think that at least for facilitator all sanity checks must be done before, else it can break in the last step by already performing some other operations.